src/io/flutter/samples/FlutterSampleNotificationProvider.java

```diff
@@ -5,65 +5,155 @@
+import com.intellij.openapi.editor.Editor;
 import com.intellij.openapi.fileEditor.FileEditor;
+import com.intellij.openapi.fileEditor.TextEditor;
 import com.intellij.openapi.project.DumbAware;
 import com.intellij.openapi.project.Project;
+import com.intellij.openapi.util.Computable;
 import com.intellij.openapi.util.Key;
 import com.intellij.openapi.util.io.FileUtil;
 import com.intellij.openapi.vfs.VirtualFile;
+import com.intellij.psi.PsiDocumentManager;
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiElementVisitor;
+import com.intellij.psi.PsiFile;
+import com.intellij.ui.EditorNotificationPanel;
 import com.intellij.ui.EditorNotifications;
+import com.jetbrains.lang.dart.psi.DartClass;
 import io.flutter.sdk.FlutterSdk;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
-import javax.swing.*;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
+import java.util.regex.Pattern;
 
-public class FlutterSampleNotificationProvider extends EditorNotifications.Provider<JPanel> implements DumbAware {
-  private static final Key<JPanel> KEY = Key.create("flutter.sample");
+public class FlutterSampleNotificationProvider extends EditorNotifications.Provider<EditorNotificationPanel> implements DumbAware {
+  private static final Key<EditorNotificationPanel> KEY = Key.create("flutter.sample");
 
-  @NotNull private final Project project;
+  @NotNull final Project project;
 
   public FlutterSampleNotificationProvider(@NotNull Project project) {
     this.project = project;
   }
 
   @NotNull
   @Override
-  public Key<JPanel> getKey() {
+  public Key<EditorNotificationPanel> getKey() {
     return KEY;
   }
 
   @Nullable
   @Override
-  public JPanel createNotificationPanel(@NotNull VirtualFile file, @NotNull FileEditor editor) {
-    final List<FlutterSample> samples = getSamplesForFile(file);
-    return !samples.isEmpty() ? new FlutterSampleActionsPanel(samples, project) : null;
-  }
-
-  private FlutterSdk getSdk() {
-    return FlutterSdk.getFlutterSdk(project);
-  }
+  public EditorNotificationPanel createNotificationPanel(
+    @NotNull VirtualFile file, @NotNull FileEditor fileEditor, @NotNull Project project) {
+    if (!(fileEditor instanceof TextEditor)) {
+      return null;
+    }
 
-  private List<FlutterSample> getSamplesForFile(@NotNull VirtualFile file) {
-    final FlutterSdk sdk = getSdk();
+    final FlutterSdk sdk = FlutterSdk.getFlutterSdk(project);
     if (sdk == null) {
-      return Collections.emptyList();
+      return null;
     }
 
+    final String flutterPackagePath = FileUtil.normalize(sdk.getHomePath()) + "/packages/flutter/lib/src/";
     final String filePath = FileUtil.normalize(file.getPath());
-    final String pathSuffix = FileUtil.normalize(sdk.getHomePath()) + "/packages/flutter/";
 
+    // Only show for files in the flutter sdk.
+    if (!filePath.startsWith(flutterPackagePath)) {
```

Comment:
Given some `VirtualFile` (even before referencing the `FlutterSdk`), it should be able to bail if the file isn't in the IntelliJ `Library`.  Exiting faster is important when an editor notification is called on every file.

Reply:
I definitely agree with the performance considerations. Do you have a pointer to the additional check here - should the File be contained by the Library? Should the Project have a dart sdk Library in the LibraryTable?

Reply:
Change the first if statement to something like

`if (!(fileEditor instanceof TextEditor) && FileTypeRegistry.getInstance().isFileOfType(file, DartFileType.INSTANCE) && !file.isWritable()) {`

This should exit quickly only for Dart files that are not writable by the user (with very few exceptions this will only be Dart/Flutter library files.)

The mentioned file index checks would be checks like this:

`(!)ProjectFileIndex.getInstance(project).isInContent(file) or ProjectFileIndex.getInstance(project).isInLibrary/[Classes/Source](file)`

These would be the APIs if you needed some more generic check (such as for all languages).

I suspect though assuming that `isWritable()` returns `false` above as it should for files in a library, I doubt that the calls out to `ProjectFileIndex` would be faster than directly checking the specific library paths like you are doing here.